File: sdk/typescript/tests/invoker.transport.test.ts
import { mkdtempSync } from "node:fs";
import { createServer } from "node:http2";
import { createServer as createNetServer } from "node:net";
import { tmpdir } from "node:os";
import { join } from "node:path";

import { create } from "@bufbuild/protobuf";
import { type ServiceImpl } from "@connectrpc/connect";
import { connectNodeAdapter } from "@connectrpc/connect-node";
import { expect, test } from "bun:test";

import {
  ExchangeInvocationTokenResponseSchema,
  OperationResultSchema,
  AppInvoker as AppInvokerService,
} from "../src/internal/gen/v1/app_pb.ts";
import {
  ENV_HOST_SERVICE_SOCKET,
  ENV_HOST_SERVICE_TOKEN,
  AppInvoker,
  request,
} from "../src/index.ts";
import { removeTempDir } from "./helpers.ts";

interface IssueLookupParams {
  issue_number: number;
}

test("AppInvoker forwards invocation tokens from strings and Request objects", async () => {
  const tempDir = mkdtempSync(join(tmpdir(), "gts-plugin-invoker-"));
  const socketPath = join(tempDir, "plugin-invoker.sock");
  const previousSocket = process.env[ENV_HOST_SERVICE_SOCKET];
  const calls: Array<{
    invocationToken: string;
    app: string;
    operation: string;
    params: Record<string, unknown>;
    connection: string;
    instance: string;
    idempotencyKey: string;
  }> = [];
  const graphqlCalls: Array<{
    invocationToken: string;
    app: string;
    document: string;
    variables: Record<string, unknown>;
    connection: string;
    instance: string;
    idempotencyKey: string;
  }> = [];
  const exchanges: Array<{
    parentInvocationToken: string;
    grants: Array<{
      app: string;
      operations: string[];
      surfaces: string[];
      allOperations: boolean;
    }>;
    ttlSeconds: bigint;
  }> = [];

  const handler = connectNodeAdapter({
    grpc: true,
    grpcWeb: false,
    connect: false,
    routes(router) {
      router.service(
        AppInvokerService,
        {
          async exchangeInvocationToken(input) {
            exchanges.push({
              parentInvocationToken: input.parentInvocationToken,
              grants: input.grants.map((grant) => ({
                app: grant.app,
                operations: [...grant.operations],
                surfaces: [...grant.surfaces],
                allOperations: grant.allOperations,
              })),
              ttlSeconds: input.ttlSeconds,
            });
            return create(ExchangeInvocationTokenResponseSchema, {
              invocationToken: `${input.parentInvocationToken}:child`,
            });
          },
          async invoke(input) {
            const params = Object.fromEntries(Object.entries(input.params ?? {}));
            calls.push({
              invocationToken: input.invocationToken,
              app: input.app,
              operation: input.operation,
              params,
              connection: input.connection,
              instance: input.instance,
              idempotencyKey: input.idempotencyKey,
            });
            return create(OperationResultSchema, {
              status: 207,
              body: JSON.stringify({
                invocationToken: input.invocationToken,
                app: input.app,
                operation: input.operation,
                params,
                connection: input.connection,
                instance: input.instance,
                idempotencyKey: input.idempotencyKey,
              }),
            });
          },
          async invokeGraphQL(input) {
            const variables = Object.fromEntries(Object.entries(input.variables ?? {}));
            graphqlCalls.push({
              invocationToken: input.invocationToken,
              app: input.app,
              document: input.document,
              variables,
              connection: input.connection,
              instance: input.instance,
              idempotencyKey: input.idempotencyKey,
            });
            return create(OperationResultSchema, {
              status: 208,
              body: JSON.stringify({
                invocationToken: input.invocationToken,
                app: input.app,
                document: input.document,
                variables,
                connection: input.connection,
                instance: input.instance,
                idempotencyKey: input.idempotencyKey,
              }),
            });
          },
        } satisfies Partial<ServiceImpl<typeof AppInvokerService>>,
      );
    },
  });
  const server = createServer(handler);

  try {
    await new Promise<void>((resolve, reject) => {
      server.once("error", reject);
      server.listen(socketPath, () => {
        server.off("error", reject);
        resolve();
      });
    });

    process.env[ENV_HOST_SERVICE_SOCKET] = socketPath;

    const fromHandle = new AppInvoker("invocation-token-123");
    const childToken = await fromHandle.exchangeInvocationToken({
      grants: [
        {
          app: "github",
          operations: ["get_issue"],
        },
        {
          app: "linear",
          surfaces: ["graphql"],
        },
        {
          app: "google_sheets",
          allOperations: true,
        },
      ],
      ttlSeconds: 45,
    });
    expect(childToken).toBe("invocation-token-123:child");

    const issueLookupParams: IssueLookupParams = {
      issue_number: 42,
    };
    const first = await fromHandle.invoke(
      "github",
      "get_issue",
      issueLookupParams,
      {
        connection: "work",
        instance: "secondary",
        idempotencyKey: " issue-42-create ",
      },
    );

    expect(first.status).toBe(207);
    expect(JSON.parse(first.body)).toEqual({
      invocationToken: "invocation-token-123",
      app: "github",
      operation: "get_issue",
      params: {
        issue_number: 42,
      },
      connection: "work",
      instance: "secondary",
      idempotencyKey: "issue-42-create",
    });

    const fromRequest = new AppInvoker(
      request("tok", {}, {}, {}, {}, {}, "invocation-token-456"),
    );
    const second = await fromRequest.invoke("slack", "post_message", {
      channel: "eng",
      text: "hello",
      nested: Object.freeze({ urgent: true }),
    });

    expect(second.status).toBe(207);
    expect(JSON.parse(second.body)).toEqual({
      invocationToken: "invocation-token-456",
      app: "slack",
      operation: "post_message",
      params: {
        channel: "eng",
        text: "hello",
        nested: { urgent: true },
      },
      connection: "",
      instance: "",
      idempotencyKey: "",
    });

    await expect(
      fromRequest.invoke("slack", "bad", { when: new Date() } as any),
    ).rejects.toThrow(TypeError);

    const graphql = await fromRequest.invokeGraphQL(
      "linear",
      "query Viewer($team: String!) { viewer(team: $team) { id } }",
      {
        variables: { team: "eng" },
        connection: "workspace",
        idempotencyKey: " graphql-call-42 ",
      },
    );

    expect(graphql.status).toBe(208);
    expect(JSON.parse(graphql.body)).toEqual({
      invocationToken: "invocation-token-456",
      app: "linear",
      document: "query Viewer($team: String!) { viewer(team: $team) { id } }",
      variables: {
        team: "eng",
      },
      connection: "workspace",
      instance: "",
      idempotencyKey: "graphql-call-42",
    });

    expect(exchanges).toEqual([
      {
        parentInvocationToken: "invocation-token-123",
        grants: [
          {
            app: "github",
            operations: ["get_issue"],
            surfaces: [],
            allOperations: false,
          },
          {
            app: "linear",
            operations: [],
            surfaces: ["graphql"],
            allOperations: false,
          },
          {
            app: "google_sheets",
            operations: [],
            surfaces: [],
            allOperations: true,
          },
        ],
        ttlSeconds: 45n,
      },
    ]);

    expect(calls).toEqual([
      {
        invocationToken: "invocation-token-123",
        app: "github",
        operation: "get_issue",
        params: {
          issue_number: 42,
        },
        connection: "work",
        instance: "secondary",
        idempotencyKey: "issue-42-create",
      },
      {
        invocationToken: "invocation-token-456",
        app: "slack",
        operation: "post_message",
        params: {
          channel: "eng",
          text: "hello",
          nested: { urgent: true },
        },
        connection: "",
        instance: "",
        idempotencyKey: "",
      },
    ]);

    expect(graphqlCalls).toEqual([
      {
        invocationToken: "invocation-token-456",
        app: "linear",
        document: "query Viewer($team: String!) { viewer(team: $team) { id } }",
        variables: {
          team: "eng",
        },
        connection: "workspace",
        instance: "",
        idempotencyKey: "graphql-call-42",
      },
    ]);
  } finally {
    if (previousSocket === undefined) {
      delete process.env[ENV_HOST_SERVICE_SOCKET];
    } else {
      process.env[ENV_HOST_SERVICE_SOCKET] = previousSocket;
    }
    if (server.listening) {
      await new Promise<void>((resolve) => {
        server.close(() => resolve());
      });
    }
    removeTempDir(tempDir);
  }
});

test("AppInvoker prioritizes invocation-token validation over socket configuration", () => {
  const previousSocket = process.env[ENV_HOST_SERVICE_SOCKET];

  try {
    delete process.env[ENV_HOST_SERVICE_SOCKET];
    expect(() => new AppInvoker("   ")).toThrow(
      "app invoker: invocation token is not available",
    );
  } finally {
    if (previousSocket === undefined) {
      delete process.env[ENV_HOST_SERVICE_SOCKET];
    } else {
      process.env[ENV_HOST_SERVICE_SOCKET] = previousSocket;
    }
  }
});

async function reserveTCPAddress(): Promise<string> {
  return await new Promise((resolve, reject) => {
    const server = createNetServer();
    server.once("error", reject);
    server.listen(0, "127.0.0.1", () => {
      const address = server.address();
      if (!address || typeof address === "string") {
        server.close();
        reject(new Error("failed to reserve tcp address"));
        return;
      }
      const result = `${address.address}:${address.port}`;
      server.close((err) => {
        if (err) {
          reject(err);
          return;
        }
        resolve(result);
      });
    });
  });
}

test("AppInvoker honors tcp target env and relay token env", async () => {
  const previousSocket = process.env[ENV_HOST_SERVICE_SOCKET];
  const previousToken = process.env[ENV_HOST_SERVICE_TOKEN];
  const seenTokens: string[] = [];
  const address = await reserveTCPAddress();

  const handler = connectNodeAdapter({
    grpc: true,
    grpcWeb: false,
    connect: false,
    routes(router) {
      router.service(AppInvokerService, {
        async invoke(input) {
          return create(OperationResultSchema, {
            status: 204,
            body: JSON.stringify({
              invocationToken: input.invocationToken,
              app: input.app,
              operation: input.operation,
            }),
          });
        },
      } satisfies Partial<ServiceImpl<typeof AppInvokerService>>);
    },
  });
  const server = createServer((req, res) => {
    const tokenHeader = req.headers["x-gestalt-host-service-relay-token"];
    if (typeof tokenHeader === "string") {
      seenTokens.push(tokenHeader);
    }
    handler(req, res);
  });

  try {
    await new Promise<void>((resolve, reject) => {
      server.once("error", reject);
      server.listen(Number(address.split(":").at(-1)), "127.0.0.1", () => {
        server.off("error", reject);
        resolve();
      });
    });

    process.env[ENV_HOST_SERVICE_SOCKET] = `tcp://${address}`;
    process.env[ENV_HOST_SERVICE_TOKEN] = "relay-token-typescript";

    const invoker = new AppInvoker("invoke-token");
    const response = await invoker.invoke("github", "get_issue");

    expect(response.status).toBe(204);
    expect(JSON.parse(response.body)).toEqual({
      invocationToken: "invoke-token",
      app: "github",
      operation: "get_issue",
    });
    expect(seenTokens).toEqual(["relay-token-typescript"]);
  } finally {
    if (previousSocket === undefined) {
      delete process.env[ENV_HOST_SERVICE_SOCKET];
    } else {
      process.env[ENV_HOST_SERVICE_SOCKET] = previousSocket;
    }
    if (previousToken === undefined) {
      delete process.env[ENV_HOST_SERVICE_TOKEN];
    } else {
      process.env[ENV_HOST_SERVICE_TOKEN] = previousToken;
    }
    if (server.listening) {
      await new Promise<void>((resolve) => {
        server.close(() => resolve());
      });
    }
  }
});

File: sdk/go/invoker_transport_test.go
package gestalt_test

import (
	"context"
	"net"
	"sync"
	"testing"

	proto "github.com/valon-technologies/gestalt/internal/gen/v1"
	gestalt "github.com/valon-technologies/gestalt/sdk/go"
	"google.golang.org/grpc"
	"google.golang.org/grpc/metadata"
	gproto "google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/structpb"
)

type pluginInvokerTransportHarness struct {
	proto.UnimplementedPluginInvokerServer

	mu       sync.Mutex
	requests []*proto.PluginInvokeRequest
	graphQL  []*proto.PluginInvokeGraphQLRequest
	tokens   []string
}

func (h *pluginInvokerTransportHarness) Invoke(ctx context.Context, req *proto.PluginInvokeRequest) (*proto.OperationResult, error) {
	md, _ := metadata.FromIncomingContext(ctx)

	h.mu.Lock()
	if values := md.Get("x-gestalt-host-service-relay-token"); len(values) > 0 {
		h.tokens = append(h.tokens, values...)
	}
	h.requests = append(h.requests, &proto.PluginInvokeRequest{
		InvocationToken: req.GetInvocationToken(),
		Plugin:          req.GetPlugin(),
		Operation:       req.GetOperation(),
		Params:          cloneStruct(req.GetParams()),
		Connection:      req.GetConnection(),
		Instance:        req.GetInstance(),
		IdempotencyKey:  req.GetIdempotencyKey(),
	})
	h.mu.Unlock()

	return &proto.OperationResult{Status: 207, Body: "relay-ok"}, nil
}

func (h *pluginInvokerTransportHarness) InvokeGraphQL(ctx context.Context, req *proto.PluginInvokeGraphQLRequest) (*proto.OperationResult, error) {
	md, _ := metadata.FromIncomingContext(ctx)

	h.mu.Lock()
	if values := md.Get("x-gestalt-host-service-relay-token"); len(values) > 0 {
		h.tokens = append(h.tokens, values...)
	}
	h.graphQL = append(h.graphQL, &proto.PluginInvokeGraphQLRequest{
		InvocationToken: req.GetInvocationToken(),
		Plugin:          req.GetPlugin(),
		Document:        req.GetDocument(),
		Variables:       cloneStruct(req.GetVariables()),
		Connection:      req.GetConnection(),
		Instance:        req.GetInstance(),
		IdempotencyKey:  req.GetIdempotencyKey(),
	})
	h.mu.Unlock()

	return &proto.OperationResult{Status: 208, Body: "graphql-ok"}, nil
}

func cloneStruct(src *structpb.Struct) *structpb.Struct {
	if src == nil {
		return nil
	}
	return gproto.Clone(src).(*structpb.Struct)
}

func TestTransport_PluginInvokerTCPTargetTokenEnv(t *testing.T) {
	address := reserveTCPAddress()
	lis, err := net.Listen("tcp", address)
	if err != nil {
		t.Fatalf("net.Listen: %v", err)
	}
	t.Cleanup(func() { _ = lis.Close() })

	harness := &pluginInvokerTransportHarness{}
	srv := grpc.NewServer()
	proto.RegisterPluginInvokerServer(srv, harness)
	go func() {
		_ = srv.Serve(lis)
	}()
	t.Cleanup(srv.Stop)

	t.Setenv(gestalt.EnvPluginInvokerSocket, "tcp://"+address)
	t.Setenv(gestalt.EnvPluginInvokerSocketToken, "relay-token-go")
	t.Setenv("http_proxy", "http://127.0.0.1:1")
	t.Setenv("https_proxy", "http://127.0.0.1:1")
	t.Setenv("HTTP_PROXY", "http://127.0.0.1:1")
	t.Setenv("HTTPS_PROXY", "http://127.0.0.1:1")

	client, err := gestalt.Invoker("parent-token")
	if err != nil {
		t.Fatalf("Invoker: %v", err)
	}
	defer func() { _ = client.Close() }()

	result, err := client.Invoke(context.Background(), "github", "get_issue", map[string]any{
		"issue_number": 42,
	}, &gestalt.InvokeOptions{
		IdempotencyKey: " issue-42-create ",
	})
	if err != nil {
		t.Fatalf("Invoke: %v", err)
	}
	if result.Status != 207 || result.Body != "relay-ok" {
		t.Fatalf("Invoke result = %+v, want status=207 body=relay-ok", result)
	}
	graphQLResult, err := client.InvokeGraphQL(context.Background(), "linear", " query { viewer { id } } ", map[string]any{
		"team": "eng",
	}, &gestalt.InvokeOptions{
		IdempotencyKey: " graphql-call-42 ",
	})
	if err != nil {
		t.Fatalf("InvokeGraphQL: %v", err)
	}
	if graphQLResult.Status != 208 || graphQLResult.Body != "graphql-ok" {
		t.Fatalf("InvokeGraphQL result = %+v, want status=208 body=graphql-ok", graphQLResult)
	}

	harness.mu.Lock()
	defer harness.mu.Unlock()
	if len(harness.tokens) != 2 || harness.tokens[0] != "relay-token-go" || harness.tokens[1] != "relay-token-go" {
		t.Fatalf("relay tokens = %#v, want two relay-token-go entries", harness.tokens)
	}
	if len(harness.requests) != 1 {
		t.Fatalf("invoke requests len = %d, want 1", len(harness.requests))
	}
	if harness.requests[0].GetInvocationToken() != "parent-token" {
		t.Fatalf("invocation token = %q, want %q", harness.requests[0].GetInvocationToken(), "parent-token")
	}
	if harness.requests[0].GetPlugin() != "github" || harness.requests[0].GetOperation() != "get_issue" {
		t.Fatalf("invoke target = %s.%s, want github.get_issue", harness.requests[0].GetPlugin(), harness.requests[0].GetOperation())
	}
	if harness.requests[0].GetIdempotencyKey() != "issue-42-create" {
		t.Fatalf("idempotency key = %q, want issue-42-create", harness.requests[0].GetIdempotencyKey())
	}
	if len(harness.graphQL) != 1 {
		t.Fatalf("graphql requests len = %d, want 1", len(harness.graphQL))
	}
	if harness.graphQL[0].GetInvocationToken() != "parent-token" {
		t.Fatalf("graphql invocation token = %q, want parent-token", harness.graphQL[0].GetInvocationToken())
	}
	if harness.graphQL[0].GetPlugin() != "linear" || harness.graphQL[0].GetDocument() != "query { viewer { id } }" {
		t.Fatalf("graphql request = %s %q, want linear trimmed document", harness.graphQL[0].GetPlugin(), harness.graphQL[0].GetDocument())
	}
	if harness.graphQL[0].GetIdempotencyKey() != "graphql-call-42" {
		t.Fatalf("graphql idempotency key = %q, want graphql-call-42", harness.graphQL[0].GetIdempotencyKey())
	}
}
